api-requests.js

```diff
@@ -0,0 +1,75 @@
+const request = require('request-promise-native');
+const moment = require('moment');
+
+// API Credentials
+const appId = '4586bfa0';
+const appKey = '4236304488967b5683897c8980725377';            
+
+function getBusArrivalJSON(postcode) {
+    
+    return request(`https://api.postcodes.io/postcodes/${postcode}`)
+
+    .then(body => {
+        const postcodeInfoAsJSON = JSON.parse(body);
+        
+        // Extract postcode lat lon
+        const latitude = postcodeInfoAsJSON['result']['latitude'];
+        const longitude = postcodeInfoAsJSON['result']['longitude'];
+        
+        const radius = 1000; 
+        const busStopURL = `https://api.tfl.gov.uk/StopPoint?stopTypes=NaptanOnstreetBusCoachStopPair&radius=${radius}&lat=${latitude}&lon=${longitude}`;   
+        return request(busStopURL);
+    })
+    
+    .catch(err => {
+        // Both 404 status rejections and invalid postcodes end up in this catch block
+        if (err.name === 'StatusCodeError') {
+            return Promise.reject('Invalid post code');
+        } else {
+            return Promise.reject('Couldn\'t reach api.postcodes.io');
+        }
+    })
+
+    .then(body => {
+        const stopPointsAsJSON = JSON.parse(body);
+        let promises = [];
+
+        // Get data from first two closest bus stops
+        stopPointsAsJSON['stopPoints'].slice(0,2).forEach(function(stop) {

```

Comment:
```javascript
const promises = Promise.all(stopPointsAsJSON['stopPoints'].slice(...).map(stop =>
  stop.lineGroup.map({ stopId } =>
    request(`https://...${stopId}...`)
  )
).flat())
```

Lots going on here:

* Map rather than foreach/push.
* `flat` to turn `[[1, 2], [3, 4]]` into `[1, 2, 3, 4]` (i.e. unwrap the nesting of the arrays).
* Use of implicit return in arrow functions rather than explicit return (does mean losing the `requestURL` variable, so it's debatable whether it's an improvement there).

I've also mixed and matched property lookup vs dot syntax. The reasoning, roughly speaking, is using lookups on the base response, but dot syntax on everything below that. It's a bit arbitrary though, so up to you (if you wanted absolute consistency, I'd personally favour dots everywhere).